### ISSUE TITLE
Entry Preview Covers UI (Wayland issues)

### ISSUE THREAD

**Issue by inventor200** [sirjuddington/SLADE#1351]
**SLADE Version**: 3.2.0
**Operating System**: Linux (Ubuntu 22.04)

**Issue Details**:
If I click on an entry (such as PLAYPAL or any map entry), the preview is at max-scale, and completely covers all of the other entries (and most of the UI for the rest of the program), preventing me from clicking on any other entry. Sometimes the rest of the uncovered UI becomes unresponsive, and I need to force-close the program, but it's still able to ask me if I want to save my work. From this, I know that it's not entirely locked up, but all buttons and functions are completely blocked by the entry preview.

This happens for entries of all types.
The previous release version worked reliably-well on my system.

**Comment by sirjuddington** [comment]
If possible, can you post a screenshot of the issue?

**Comment by inventor200** [comment]
Absolutely. This is my first time attaching a file here, so hopefully this uploads and opens correctly.

The screenshot is me opening the freedoom2.wad, and selecting the MAP01 entry. If I try to click below the preview to reach other entries, it registers as clicking on the preview still.

Happy to add any other needed information.

![Screenshot from 2022-04-23 21-41-38](https://user-images.githubusercontent.com/26094165/164952103-f50aff5f-0b39-470f-8ea8-b7e5bf738b40.png)

**Comment by sirjuddington** [comment]
Hmm I think it might be to do with the OpenGL canvas - does it happen with text entries?

How did you get SLADE - from the drd repository, the flatpak or did you build it yourself?

Also do you have fractional scaling enabled in the display settings? It might be that the wxWidgets version in your build doesn't properly support gtk fractional scaling.

**Comment by inventor200** [comment]
Text entries seem to render perfectly well. Would also like to note that the error happens with PLAYPAL entries as well, but if that's a non-text type of data, then I'm guessing it also uses an OpenGL canvas?

I believe it's from the DRD repo? I went into terminal and entered in "apt install slade", so it's a tracked package on my system for automatic updates and such. I don't remember if I had to add DRD as a package source to do that. Otherwise I'm pretty sure it's not a flatpak, and I know I didn't build it myself.

I do not have fractional scaling enabled. Display scale is at 100%.

![Screenshot from 2022-04-24 05-48-50](https://user-images.githubusercontent.com/26094165/164970623-80a2530a-a028-4c00-b46d-0bba1c3856f0.png)

**Comment by sirjuddington** [comment]
That's really weird then, it looks like it is the right size, but it's putting it at the very top left of the parent panel instead of where it should be in the layout. What happens if you resize the window?

**Comment by inventor200** [comment]
Text-based entries are totally fine, and everything resizes as expected.

![Screenshot from 2022-04-24 17-14-21](https://user-images.githubusercontent.com/26094165/164996871-7c9abd13-036a-4070-b698-b3a4832ed4b6.png)

For other entry types, such as MAP01 from freedoom2.wad, it shows the preview getting smaller (while still obstructing the other entries as before). Another weird side-effect of this, though: the actual Slade window does not change size (like it does when resizing the program with a text entry open), and now I can neither hover nor click the maximize button or the close button of the window's header bar.

After attempting this resize, the ability to pull at the edges of the window are indicated when hovering (the cursor turns into an edge-drag icon), but clicking and attempting to actually do this action cause the cursor to reset and no action is taken at all.

![Screenshot from 2022-04-24 17-14-48](https://user-images.githubusercontent.com/26094165/164996872-c7088705-9e78-4cfa-b185-5fcd51333f25.png)

_(While I am not familiar with the code base, I'm willing to bet that the window's internal model thinks that it achieved a smaller size, and is no longer maximized. However, whatever is handling the rendering for the window failed to update and redraw the window at its correct size. Probably, if I clicked enough times in the empty space where the maximize button or close button SHOULD be, I would eventually find where the internal model THINKS the buttons are positioned now, and the window would close or maximize as expected. This is just speculation, but an error like this once happened in one of my projects a few years ago. I could be totally wrong, though.)_

When I close the window (by right-clicking the Slade icon in the taskbar, and selecting "close"), it does the normal close operations ("Are you sure you want to exit Slade?" modal pops up and works perfectly). When I re-open Slade, it is now at the correct reduced size, even after the bugs and lack of apparent size change when previewing MAP01.

![Screenshot from 2022-04-24 17-32-27](https://user-images.githubusercontent.com/26094165/164997558-b9023f93-8123-4e15-8146-5c6f60290585.png)

I tried moving the divider between the entries list and the preview area after selecting MAP01 for preview. When moving to the right, the map preview gets smaller, but does not change position. However, the width of the entries list also does not change, and the divider does not visibly move while dragging. When I stop dragging, the cursor icon stays in place, despite not hovering over the visible divider anymore.

When dragging this divider with a text entry previewed, everything behaves entirely as expected.

I'm happy to perform more tests and report back as needed.

**Comment by sirjuddington** [comment]
It looks like something is going really wrong with the window manager - I just noticed you listed ubuntu 22.04 as your OS, is that correct? If so it was just released a few days ago and I'm wondering if there might be some issues between wxWidgets/OpenGL and whatever newer version of GTK/Wayland etc it's using

**Comment by inventor200** [comment]
Yeah, I'm using X11 and Gnome still; haven't switched to Wayland.

Figured it was my Linux. I have a Windows VM on it, and the Slade 3.2.0 runs fine on that. The problem is all my asset creation tools are on the Linux, so I would need to truck the project back and forth between the VM and the host, and that would be a pain in my workflow.

**Comment by sirjuddington** [comment]
Looking into it more, #1352 seems to fix the issue, though I'm not exactly sure why. I was a bit reluctant to put that in for 3.2.1 since it's a fairly major change, but might need to rethink that now

**Comment by sirjuddington** [comment]
You can just leave it for now - it looks like #1352 doesn't actually fix it, or at least not always - it's started spontaneously happening on my laptop now even with that change, so I have absolutely no idea what it could be unfortunately

**Comment by inventor200** [comment]
Alrighty; I shall leave it, then. Thank you again for all that you do!

**Comment by sirjuddington** [comment]
Ok I've found a workaround at least - GDK_BACKEND needs to be set to x11, so you will have to run slade from a terminal like:

```
GDK_BACKEND=x11 slade
```

This works for me at least, even without #1352

**Comment by inventor200** [comment]
Aha, excellent! Works perfectly on my end as well, too! Need to save this into a bash script somewhere so I remember to run it this way from now on.

**Comment by inventor200** [comment]
For anyone else who might find this thread when searching the Internet, you can implement the workaround in a .desktop file (Slade's is found in `/usr/share/applications/Slade.desktop`).

Change the exec line to the following:
`Exec=env GDK_BACKEND=x11 slade`

This will allow you to keep your taskbar icon and application list icon the same, and you can click them as you used to before when opening Slade, but now it opens it with the GDK workaround!

**Comment by axredneck** [comment]
I have this issue when using Slade 3.2.1 under Plasma Wayland session:
![Screenshot_20221024_000333](https://user-images.githubusercontent.com/20617900/197418341-223af592-63e9-4616-ab97-1c1c27165170.png)

**Comment by joanbm** [comment]
Same on Wayland+Sway.

Possibly a wxWidgets bug? There was a recent fix related to OpenGL canvas positioning on Wayland there, but I tried it and it doesn't appear to fix SLADE: https://github.com/wxWidgets/wxWidgets/commit/800d6ed69b9f6b909a36304ebddb52047c691838

I tried adding some `printf`'s to that place on wxWidgets and it seems to get the right coordinates, but the canvas stays at its initial position for some reason.

**Comment by axredneck** [comment]
Works fine for me now. Fixed?

**Comment by joanbm** [comment]
> Works fine for me now. Fixed?

Still reproduces for me on SLADE 3.2.1 on Arch Linux (updated with recent packages, e.g. GTK 3.24.36, wxWidgets 3.2.2.1, wlroots 0.16.2 + Sway 1.8.1, etc.).

**Comment by joanbm** [comment]
The problem where the entry preview is placed over the list of entries (as seen in https://github.com/sirjuddington/SLADE/issues/1351#issuecomment-1107683199) appears to be related to the "Start Page" tab.

The "Start Page" tab is a WebView, which internally creates an OpenGL context, and it seems to have some bad interaction with the OpenGL context used for the entry preview, I haven't figured out what's going on yet.

If I disable it (uncheck Edit -> Preferences -> Interface -> "Show Start Page on Startup" and restart), the entry preview is properly placed.

However, even with that workaround, it's still unusable due to another issue: https://github.com/wxWidgets/wxWidgets/issues/22580 where the entry preview canvas doesn't hide after changing to another entry.

**Comment by joanbm** [comment]
SLADE should work on native Wayland under wxWidgets 3.2.3 (which includes [this fix](https://github.com/wxWidgets/wxWidgets/pull/23835)) as long as you disable "Show Start Page on Startup" on Preferences and restart.

The problem with the "Start Page" is a GTK3 issue, related to [GTK issue 5392](https://gitlab.gnome.org/GNOME/gtk/-/issues/5392) and [fixed in GTK4](https://gitlab.gnome.org/GNOME/gtk/-/commit/c6cef6db526930b4cd925c0998da5acc4d4ab35d) but not GTK3. I have a backport in the works but I haven't been able to submit it there yet.

**Comment by Blzut3** [comment]
Can confirm that without the start page wxWidgets 3.2.3 is working under Wayland.  It's unfortunate there's a GTK bug too since it probably means it will be some time before I can enable native Wayland in the deb package even if a backport goes in upstream.

**Comment by SteelT1** [comment]
Disabling the "Start Page" seems to also fix the entire UI no longer updating when attempting to preview any graphics lumps, or double clicking on a `TEXTURES` lump to open the TEXTUREx Editor, under Wayland.

**Comment by joanbm** [comment]
I opened [a MR on GTK3's repo](https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8895) that should fix this bug (the glitch that causes the interference across the Start Page's WebView and the other OpenGL canvases from SLADE).

**Comment by joanbm** [comment]
The fix above has been released in GTK 3.24.51, if you are in a distro that already has it (e.g. Arch Linux), I'd be grateful if you can test it and report if it fixes this & other issues in Wayland.

**Comment by Blzut3** [comment]
@joanbm While my distro doesn't provide GTK 3.24.51, I did compile it and load it via LD_LIBRARY_PATH.  Verified that proc maps shows the correct so file being loaded:
```
blzut3@AMD-9800X3D:~/Code/Public/Slade/build-wx33$ grep gtk /proc/$(pgrep slade)/maps
7a33e8000000-7a33e820f000 r--p 00000000 103:07 26097469                  /home/blzut3/Code/Stock/gtk-3.24.51/build/gtk/libgtk-3.so.0.2419.32
7a33e820f000-7a33e85b8000 r-xp 0020f000 103:07 26097469                  /home/blzut3/Code/Stock/gtk-3.24.51/build/gtk/libgtk-3.so.0.2419.32
7a33e85b8000-7a33e8734000 r--p 005b8000 103:07 26097469                  /home/blzut3/Code/Stock/gtk-3.24.51/build/gtk/libgtk-3.so.0.2419.32
7a33e8734000-7a33e8747000 r--p 00733000 103:07 26097469                  /home/blzut3/Code/Stock/gtk-3.24.51/build/gtk/libgtk-3.so.0.2419.32
7a33e8747000-7a33e874a000 rw-p 00746000 103:07 26097469                  /home/blzut3/Code/Stock/gtk-3.24.51/build/gtk/libgtk-3.so.0.2419.32
7a33e8800000-7a33e895a000 r--p 00000000 103:04 9962187                   /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0.8.9
7a33e895a000-7a33ea020000 r-xp 0015a000 103:04 9962187                   /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0.8.9
7a33ea020000-7a33ea43e000 r--p 01820000 103:04 9962187                   /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0.8.9
7a33ea43e000-7a33ea4a1000 r--p 01c3d000 103:04 9962187                   /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0.8.9
7a33ea4a1000-7a33ea4b7000 rw-p 01ca0000 103:04 9962187                   /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0.8.9
7a33ea600000-7a33ead21000 r--p 00000000 103:04 9963141                   /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.1.so.0.17.11
7a33ead21000-7a33ee3b5000 r-xp 00721000 103:04 9963141                   /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.1.so.0.17.11
7a33ee3b5000-7a33ef7c7000 r--p 03db5000 103:04 9963141                   /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.1.so.0.17.11
7a33ef7c7000-7a33efaaa000 r--p 051c7000 103:04 9963141                   /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.1.so.0.17.11
7a33efaaa000-7a33efaed000 rw-p 054aa000 103:04 9963141                   /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.1.so.0.17.11
```
Does not seem to change anything for me vs the Ubuntu 25.10 provided gtk 3 (3.24.50).  Built with statically linked wxWidgets 3.3.1.

**Comment by Blzut3** [comment]
@joanbm Scratch my (now deleted) reply.  Saw your changes were in g**d**k and made sure to load that (same gtk 3.24.51 source tarball).  It does seem to fix this issue.

There does seem to still be some conflict between `wxUSE_GLCANVAS_EGL` and X11.  This is required to be on for Wayland, but causes Slade to crash opening the texture editor if enabled and running under X11.  Of course we're tracking the texture editor issues separate from this ticket.  Package maintainers should probably continue to refrain from enabling Wayland until those are resolved due to this mutual exclusivity.  But users compiling themselves should probably experiment with Wayland if they have the requisite version of GTK/GDK.

**Comment by Blzut3** [comment]
Noting here as well that pull request #1863 should fix the problem with X11+EGL, so I think at this point the only thing we're waiting on is a wxWidgets 3.3.2 release for pointer warping to work and I think once that's confirmed that'll solve the known issues.

As mentioned earlier the deb packages won't be able to enable native Wayland support for awhile due to (attempting to anyway) support Ubuntu LTS (and ideally Debian) for the full 5 year cycle.  It's unlikely the GDK bug fix will be back ported*, so will just have to live with XWayland until 2029/2030.  Other maintainers won't be as restricted.

\* Yes, I do find it kind of silly that Debian and Ubuntu don't pull the fixes on libraries that are in pure maintenance mode upstream, but I get under normal circumstances everyone's definition of "maintenance release" differs so it's hard to make a policy around that.
